apis/Google.Cloud.Firestore.Data/Google.Cloud.Firestore.Data/ValueDeserializer.cs

```diff
@@ -141,28 +140,35 @@ internal object Deserialize(FirestoreDb db, Value value, BclType targetType)
 
         private object DeserializeMap(FirestoreDb db, IDictionary<string, Value> values, BclType targetType)
         {
-            // TODO: Handle dictionaries where TValue isn't object
+            // TODO: Support deserializatoin to IReadOnlyDictionary<,> as well? (It becomes somewhat awkward, for limited benefit.)
             if (targetType == typeof(object) || targetType == typeof(IDictionary<string, object>))
             {
                 targetType = _defaultMapType;
             }
-            if (targetType == typeof(Dictionary<string, object>))
+
+            // Shortcut to avoid reflection
+            if (targetType == typeof(Dictionary<string, object>) || targetType == typeof(ExpandoObject))
             {
-                var ret = new Dictionary<string, object>();
-                foreach (var pair in values)
-                {
-                    ret[pair.Key] = Deserialize(db, pair.Value, typeof(object));
-                }
-                return ret;
+                return PopulateDictionary<object>(db, values, targetType);
             }
-            if (targetType == typeof(ExpandoObject))
+            if (TryGetStringDictionaryValueType(targetType, out var dictionaryElementType))
             {
-                IDictionary<string, object> ret = new ExpandoObject();
-                foreach (var pair in values)
+                // If we've been asked to deserialize to an interface, let's just see if Dictionary<,>
+                // would work, but otherwise give up.
+                if (targetType.GetTypeInfo().IsInterface)
                 {
-                    ret[pair.Key] = Deserialize(db, pair.Value, typeof(object));
+                    var candidateType = typeof(Dictionary<,>).MakeGenericType(typeof(string), dictionaryElementType);
+                    if (targetType.IsAssignableFrom(candidateType))
+                    {
+                        targetType = candidateType;
+                    }
+                    else
+                    {
+                        throw new ArgumentException($"INTERFACE Unable to deserialize map to {targetType}");
```

Comment:
Whoops, just realized this message needs changing (the INTERFACE part was for debugging). Need to fix that and add a test...